core/backend/src/IModelDb.ts

```diff
@@ -82,14 +83,93 @@ export interface ComputedProjectExtents {
   outliers?: Id64Array;
 }
 
+/**
+ * @beta
+ */
+export interface LockControl {
+  /**
+   * true if this LockControl uses a server-based concurrency approach.
+   */
+  readonly isServerBased: boolean;
+  /**
+   * Close the local lock control database
+   * @internal
+   */
+  close(): void;
+  /**
+   * Notification that a new element was just created. Called by [[Element.onInserted]]
+   * @internal
+   */
+  elementWasCreated(id: Id64String): void;
+  /**
+   * Throw if locks are required and the exclusive lock is not held on the supplied element.
+   * Note: there is no need to check the shared locks on parents/models since an element cannot hold the exclusive lock without first obtaining them.
+   * Called by [[Element.onUpdate]], [[Element.onDelete]], etc.
+   * @internal
+   */
+  checkExclusiveLock(id: Id64String, type: string, operation: string): void;
+  /**
+   * Throw if locks are required and a shared lock is not held on the supplied element.
+   * Called by [[Element.onInsert]] to ensure shared lock is held on model and parent.
+   * @internal
+   */
+  checkSharedLock(id: Id64String, type: string, operation: string): void;
+  /**
+   * Determine whether the supplied element currently holds the exclusive lock
+   */
+  holdsExclusiveLock(id: Id64String): boolean;

```

Comment:
I'm not sure where `holdsXXLock` will be used. It is currently only needed for tests, but I suspect it will be useful for tools.

Reply:
>  If so, is it potentially weird that NoLocks.checkExclusive/SharedLock never throw but NoLocks.holdsExclusive/SharedLock always return false?

That's why I made `checkXXXLock` @internal - because that can be somewhat confusing. Note that the comment says "Throw **if locks are required and**..."

Reply:
What would be more useful is an API that allows the app to query whether a lock is *available* - i.e., not held by anyone else. For example, when mousing over an element with a modification tool, the tool would want to provide feedback as to whether it will be able to operate on that element. It won't want to acquire the lock(s) yet, only query whether it would be able to if it needed them.

Reply:
Maybe, but when would a tool use that? If you're about to modify an element you should attempt to acquire its lock, rather than query and then acquire (2 round trips). Plus, just because a lock is available in the query doesn't mean it will still be available when you attempt to lock it. Also, I don't know if the current IMH lock api has that, if we decide we should add it.

Reply:
When I mouse over an element I'm not necessarily about to modify it, but am potentially interested in whether I could modify it. We had this in bim02. It required downloading and caching information about *all* locks held by other people. This was expensive and of course the cache could get out of date. But we implemented it because we believed tools needed it. Maybe they don't.

Reply:
The IMH team objected strongly to the query that got info about all locks. We might be OK if we just queried a single lock. Maybe we could accumulate individual query results in a cache for a limited time to avoid too many individual queries.

Reply:
I know you can query for all *your* locks, but can you query about other users' locks?